controllers/s3utils.go

```diff
@@ -0,0 +1,404 @@
+func (s *s3Connection) downloadTypedObjects(bucket string,
+	objectType reflect.Type) (interface{}, error) {
+	keyPrefix := objectType.String() + "/"
+
+	keys, err := s.listKeys(bucket, keyPrefix)
+	if err != nil {
+		log.Infof("unable to listKeys of type %v from endpoint %v bucket %v keyPrefix %v err %v",
+			objectType, s.endpoint, bucket, keyPrefix, err)
+
+		return nil, err
+	}
+
+	objects := reflect.MakeSlice(reflect.SliceOf(objectType),
+		len(keys), len(keys))
+
+	for i := range keys {
+		objectReceiver := objects.Index(i).Addr().Interface()
+		if err := s.downloadObject(bucket, keys[i], objectReceiver); err != nil {
+			log.Infof("unable to downloadObject endpoint %v bucket %v key %v err %v",
+				s.endpoint, bucket, keys[i], err)
+
+			return nil, err
+		}
+	}
+
+	// Return []objectType
+	return objects.Interface(), nil
+}
+
+// List the keys (of objects) with the given keyPrefix in the given bucket.
+// - If bucket doesn't exists, will return ErrCodeNoSuchBucket "NoSuchBucket"
+// - Refer to aws documentation of s3.ListObjectsV2Input for more list options
+func (s *s3Connection) listKeys(bucket string, keyPrefix string) (
+	keys []string, err error) {
+	var nextContinuationToken *string
+
+	for gotAllObjects := false; !gotAllObjects; {
+		result, err := s.client.ListObjectsV2(&s3.ListObjectsV2Input{
+			Bucket:            &bucket,
+			Prefix:            &keyPrefix,
+			ContinuationToken: nextContinuationToken,
+		})
+		if err != nil {
+			return nil,
+				errorswrapper.Wrapf(err, "Failed to list objects in bucket %s:%s",
+					bucket, keyPrefix)
+		}
+
+		for _, entry := range result.Contents {
+			keys = append(keys, *entry.Key)
+		}
+
+		if *result.IsTruncated {
+			nextContinuationToken = result.NextContinuationToken
+		} else {
+			gotAllObjects = true
+		}
+	}
+
+	return
+}
+
+// Download the given object from the given bucket with the given key
+// - OK to call DownloadObject() concurrently from multiple goroutines safely.
+// - Only those type field name in the downloaded json blob that are also present
+// 	 in the downloadContent type will be filled; other fields will be dropped
+//	 without returning any error.  More info at documentation of json.Unmarshall()
+// - Download may fail due to many reasons: RequestError (connection error),
+//	 NoSuchBucket, NoSuchKey, invalid gzip header, json unmarshall error,
+//	 InvalidParameter (e.g., empty key), etc.
+// - Assumes that the downloaded objects are json blobs that have been then gzipped
+//	 and hence, will attempt to unzip and decode the json blobs
+func (s *s3Connection) downloadObject(bucket string, key string,
+	downloadContent interface{}) error {
+	writerAt := &aws.WriteAtBuffer{}
+	if _, err := s.downloader.Download(writerAt, &s3.GetObjectInput{
+		Bucket: &bucket,
+		Key:    &key,
+	}); err != nil {
+		return errorswrapper.Wrapf(err,
+			"Failed to download data of %s:%s", bucket, key)
+	}
+
+	gzReader, err := gzip.NewReader(bytes.NewReader(writerAt.Bytes()))
```

Comment:
same `defer` here

Reply:
Same response: given that I want to return an err if close fails,  I need to defer a func() to change the return value.  Is that what you mean?